File: CCode/PlayerMovement.cs
using System;
using UnityEngine;

[Serializable]
public class PlayerMovement : MonoBehaviour
{
	public Item FeetEquipment;
	public float runSpeed;

	public float accel;

	public float jump_force;

	public float gravity;

	public bool slow_fall;

	public Animator _animator;

	public SpriteRenderer _sprite;

	public bool flipX;

	private HandScript hand;

	public Vector2 externalForces;

	public float horizontalInput;

	public float verticalInput;

	public RaycastHit2D floor_ray;

	public RaycastHit2D left_ray;

	public RaycastHit2D right_ray;

	public Vector2 staticFrame;

	public Vector2 currSpeed;

	public bool isStillJumping;

	public bool isAirborn;

	public bool isOnLeftWall;

	public bool isOnRightWall;

	public bool isWallJumping;

	public float timeTouchingSomething;

	public float timeTouchingTheFloor;

	public bool isTouchingFloor;

	public bool isTouchingSomething;

	public float speedMax;

	public Rigidbody2D rigid;

	public BoxCollider2D head;

	public InputManager Inputs;

	public LayerMask floor_layer_mask;

	public int ground_speed;

	public int player_num;

	public int health;

	public GameManager gameManager;
    public bool can_jump;
    public bool jump_triggered;
    public float slow_fall_speed;

    public PlayerMovement()
	{
		this.runSpeed = 35f;
		this.accel = (float)1;
		this.jump_force = (float)35;
		this.gravity = 65f;
		this.floor_ray = default(RaycastHit2D);
		this.left_ray = default(RaycastHit2D);
		this.right_ray = default(RaycastHit2D);
	}

	public  void Start()
	{
		this.Inputs = (InputManager)this.GetComponent("InputManager");
		Time.timeScale = (float)1;
		float time = Time.time;
		this.rigid = (Rigidbody2D)this.GetComponent(typeof(Rigidbody2D));
		this.hand = (HandScript)this.gameObject.transform.Find("Hand").GetComponent("HandScript");
	}

	public  void Update()
	{
		int num = 0;
		Vector3 position = this.transform.position;
		float num2 = position.z = (float)num;
		Vector3 vector = this.transform.position = position;
		this.currSpeed = this.rigid.velocity;
		this.currSpeed += this.externalForces;
		this.externalForces = default(Vector2);
		this.getInputs();
		this.whatAmITouching();
		if (this.Inputs.down.state == key_state.Hold)
		{
			this.head.gameObject.SetActive(false);
		}
		else
		{
			this.head.gameObject.SetActive(true);
		}
		this.handleHorizontal();
		this.handleJumps();
		this.applyGravity();
		this.rigid.velocity = this.currSpeed;
		this.UpdateAnimator();
		this.setStaticFrame(default(Vector2));
    }

	public  void UpdateAnimator()
	{
		if (this.horizontalInput != (float)0)
		{
			if (this.horizontalInput < (float)0)
			{
				this.flipX = true;
			}
			else
			{
				this.flipX = false;
			}
		}
		this._animator.SetBool("Grounded", this.isTouchingFloor);
		this._animator.SetFloat("Speed", Mathf.Abs(this.currSpeed.x) * (float)20);
		this._animator.SetFloat("vSpeed", this.currSpeed.y);
		this._animator.SetBool("Crouching", this.Inputs.down.state == key_state.Hold);
		this._animator.SetBool("LookingUp", this.Inputs.up.state == key_state.Hold);
			_animator.SetBool("WallClinging", !isTouchingFloor && (isOnLeftWall || isOnRightWall));

		
		this._animator.SetBool("Jetpacking", this.slow_fall);
		if (this.isOnRightWall)
		{
			this.flipX = true;
		}
		if (this.isOnLeftWall)
		{
			this.flipX = false;
		}
		if (this.flipX)
		{
			int num = -1;
			Vector3 localScale = this.transform.localScale;
			float num2 = localScale.x = (float)num;
			Vector3 vector = this.transform.localScale = localScale;
		}
		else
		{
			int num3 = 1;
			Vector3 localScale2 = this.transform.localScale;
			float num4 = localScale2.x = (float)num3;
			Vector3 vector2 = this.transform.localScale = localScale2;
		}
	}

	public  void handleJumps()
	{
		float num = this.jump_force;
        if (isTouchingFloor) can_jump = true;
        if (this.Inputs.jump.state == key_state.Down) jump_triggered = true;
		if ( jump_triggered)
		{
            jump_triggered = false;
			if (can_jump)
			{
				this.isStillJumping = true;
				this.currSpeed.y = this.currSpeed.y + (this.staticFrame.y + num);
			}
			else if (this.isOnLeftWall || this.isOnRightWall)
			{
				this.currSpeed.y = this.staticFrame.y + num / 1.5f;
				this.currSpeed.x = this.currSpeed.x + ((!this.isOnLeftWall) ? ((float)0) : num);
				this.currSpeed.x = this.currSpeed.x - ((!this.isOnRightWall) ? ((float)0) : num);
				this.isStillJumping = true;
			}
            
		}
		if (this.Inputs.jump.state == key_state.Up)
		{
			if (this.isStillJumping && this.currSpeed.y > (float)5)
			{
				this.currSpeed.y = this.currSpeed.y - Mathf.Min(this.currSpeed.y / (float)2, num / (float)2);
			}
			this.isStillJumping = false;
		}
		bool arg_169_0;
		if (arg_169_0 = !this.isTouchingFloor)
		{
			arg_169_0 = !this.isStillJumping;
		}
		bool arg_183_1;
		if (arg_183_1 = arg_169_0)
		{
			arg_183_1 = (this.Inputs.jump.state == key_state.Hold);
		}
        if (!isTouchingFloor || isStillJumping) can_jump = false;
    }

	public  void handleHorizontal()
	{
		this.ground_speed = (int)(this.currSpeed.x - this.staticFrame.x);
		float ground_speed_magnitude = Mathf.Abs(this.ground_speed);
		this.speedMax = (this.Inputs.down.state == key_state.None)?this.runSpeed:this.runSpeed/3 + Mathf.Abs(this.staticFrame.x);
		float num2 = this.speedMax - (float)ground_speed_magnitude;
		if (Mathf.Abs(this.horizontalInput) > 0.1f)
		{
			if (Mathf.Sign(this.horizontalInput) == Mathf.Sign((float)this.ground_speed))
			{
                if (num2 < 0) num2 *= 3;
				this.currSpeed.x = this.currSpeed.x + num2 * (Time.fixedDeltaTime * this.accel) * Mathf.Sign(this.horizontalInput);
			}
			else
			{
				this.currSpeed.x = this.currSpeed.x + this.speedMax * (Time.fixedDeltaTime * this.accel) * Mathf.Sign(this.horizontalInput) * (float)2;
			}
		}
		else if (ground_speed_magnitude > 4)
		{
			if (this.Inputs.down.state == key_state.Hold)
			{
				this.currSpeed.x = this.currSpeed.x - (float)this.ground_speed * Time.fixedDeltaTime * this.accel / 1.5f;
			}
			else
			{
				this.currSpeed.x = this.currSpeed.x - (float)this.ground_speed * Time.fixedDeltaTime * this.accel * (float)4;
			}
		}
		else
		{
			this.currSpeed.x = this.staticFrame.x;
		}
	}

	public  void getInputs()
	{
		if (this.Inputs.left.state == key_state.Hold)
		{
			this.horizontalInput = (float)-1;
		}
		else if (this.Inputs.right.state == key_state.Hold)
		{
			this.horizontalInput = (float)1;
		}
		else
		{
			this.horizontalInput = (float)0;
		}
		if (this.Inputs.down.state == key_state.Hold)
		{
			this.verticalInput = (float)-1;
		}
		else if (this.Inputs.up.state == key_state.Hold)
		{
			this.verticalInput = (float)1;
		}
		else
		{
			this.verticalInput = (float)0;
		}
	}

	public  void whatAmITouching()
	{
		this.isTouchingFloor = false;
		this.isOnLeftWall = false;
		this.isOnRightWall = false;
		int num = 1;
		Vector2 vector = new Vector2(this.transform.position.x, this.transform.position.y);
		Vector2 vector2 = new Vector2((this.transform.position - this.transform.right * (float)num).x, (this.transform.position - this.transform.right * (float)num).y);
		Vector2 vector3 = new Vector2((this.transform.position + this.transform.right * (float)num).x, (this.transform.position + this.transform.right * (float)num).y);
		Vector2 a = new Vector2((this.transform.position - this.transform.up * 1.5f).x, (this.transform.position - this.transform.up * 1.5f).y);
		this.right_ray = Physics2D.Linecast(vector, vector3, this.floor_layer_mask);
		if (this.right_ray.collider != null)
		{
			this.isOnRightWall = true;
			Debug.DrawLine(vector, vector3, Color.red, (float)2);
		}
		this.right_ray = Physics2D.Linecast(vector + new Vector2((float)0, (float)1), vector3 + new Vector2((float)0, (float)1), this.floor_layer_mask);
		if (this.right_ray.collider != null)
		{
			this.isOnRightWall = true;
			Debug.DrawLine(vector + new Vector2((float)0, (float)1), vector3 + new Vector2((float)0, (float)1), Color.red, (float)2);
		}
		this.left_ray = Physics2D.Linecast(vector + new Vector2((float)0, (float)1), vector2 + new Vector2((float)0, (float)1), this.floor_layer_mask);
		if (this.left_ray.collider != null)
		{
			this.isOnLeftWall = true;
			Debug.DrawLine(vector + new Vector2((float)0, (float)1), vector2 + new Vector2((float)0, (float)1), Color.yellow, (float)2);
		}
		this.left_ray = Physics2D.Linecast(vector, vector2, this.floor_layer_mask);
		if (this.left_ray.collider != null)
		{
			this.isOnLeftWall = true;
			Debug.DrawLine(vector, vector2, Color.yellow, (float)2);
		}
		this.floor_ray = Physics2D.Linecast(vector + Vector2.left / (float)2, a + Vector2.left / (float)2, this.floor_layer_mask);
		if (this.floor_ray.collider != null)
		{
			this.isTouchingFloor = true;
			Debug.DrawLine(vector + Vector2.left / (float)2, a + Vector2.left / (float)2, Color.blue, (float)2);
		}
		this.floor_ray = Physics2D.Linecast(vector - Vector2.left / (float)2, a - Vector2.left / (float)2, this.floor_layer_mask);
		if (this.floor_ray.collider != null)
		{
			this.isTouchingFloor = true;
			Debug.DrawLine(vector - Vector2.left / (float)2, a - Vector2.left / (float)2, Color.blue, (float)2);
		}
		bool arg_44B_0;
		if (arg_44B_0 = !this.isTouchingFloor)
		{
			arg_44B_0 = !this.isOnLeftWall;
		}
		bool arg_45B_1;
		if (arg_45B_1 = arg_44B_0)
		{
			arg_45B_1 = !this.isOnRightWall;
		}
		this.isAirborn = arg_45B_1;
	}

	public  void OnCollisionStay2D(Collision2D hit)
	{
		if (hit.gameObject.GetComponent<Rigidbody2D>() && hit.rigidbody.mass * hit.rigidbody.velocity.magnitude > this.GetComponent<Rigidbody2D>().mass * this.GetComponent<Rigidbody2D>().velocity.magnitude)
		{
			this.setStaticFrame(hit.gameObject.GetComponent<Rigidbody2D>().velocity);
		}
	}

	public  void setStaticFrame(Vector3 outsideAction)
	{
	}

	public  void applyForce(Vector2 force)
	{
		this.externalForces += ((!this.flipX) ? force : new Vector2(-force.x, force.y));
	}

	public  void applyGlobalForce(Vector2 force)
	{
		this.externalForces += force;
	}

	public  void applyGravity()
	{
		if (this.currSpeed.y > (float)-140)
		{
			if (this.currSpeed.y < (float)slow_fall_speed &&( (this.isOnLeftWall && this.Inputs.left.state == key_state.Hold) || (this.isOnRightWall && this.Inputs.right.state == key_state.Hold) || (this.slow_fall)))
			{
				this.currSpeed.y = this.currSpeed.y / 1.02f;
			}
			else
			{
				this.currSpeed.y = this.currSpeed.y - this.gravity * Time.fixedDeltaTime;
			}
		}
        slow_fall = false;

    }

	public void ReceiveDamage(int amount){
		health-=amount;
		if(health<0){
			KillPlayer();
		}
	}
	public void KillPlayer(){
		gameManager.SpawnPlayer(player_num);
		Destroy(this.gameObject);
	}

	public  void Main()
	{
	}
}

File: new code/PlayerMovement.js
public var runSpeed: float = 35.0;
public var accel: float = 1;
public var jump_force: float = 35;
public var gravity: float = 65.0;
var slow_fall = false;
public var _animator: Animator;
public var _sprite: SpriteRenderer;
public var flipX: boolean;
private var hand: HandScript;
var externalForces: Vector2;
var horizontalInput: float = 0;
var verticalInput: float = 0;
public var floor_ray: RaycastHit2D = new RaycastHit2D();
public var left_ray: RaycastHit2D = new RaycastHit2D();
public var right_ray: RaycastHit2D = new RaycastHit2D();
public var staticFrame: Vector3;
public var currSpeed: Vector3;
public var isStillJumping: boolean;
public var isAirborn: boolean;
public var isOnLeftWall: boolean = false;
public var isOnRightWall: boolean = false;
public var isWallJumping: boolean = false;
public var timeTouchingSomething: float;
public var timeTouchingTheFloor: float;
public var isTouchingFloor: boolean;
public var isTouchingSomething: boolean;
public var speedMax: float = 0;
public var rigid: Rigidbody2D;
public var head: BoxCollider2D;
public var Inputs: InputManager;
public var floor_layer_mask: LayerMask;
public var gameManager : GameManager;
public var player_num : int;
public var health : int;

function Start() {
	Inputs = GetComponent("InputManager");
	Time.timeScale = 1;
	timeAtStart = Time.time;
	rigid = GetComponent(Rigidbody2D);
	hand = gameObject.transform.Find("Hand").GetComponent("HandScript");
	gameManager = GameObject.Find("GameManager").GetComponent("GameManager");
}

function Update() {

	transform.position.z = 0;
	currSpeed = rigid.velocity;  //1:Get Velocity going into this frame
	currSpeed += externalForces;
	externalForces = new Vector2();
	getInputs();//3:Get what the user is trying to do this frame
	whatAmITouching();
	if (Inputs.down.state == key_state.Hold) head.gameObject.SetActive(false);
	else head.gameObject.SetActive(true);;
	handleHorizontal(); //4:Apply LeftRight Movements
	handleJumps();//5: Apply Up down movements
	applyGravity();
	rigid.velocity = currSpeed;
	UpdateAnimator();
	setStaticFrame(new Vector2());
}
/// <summary>
/// This is called at Update() and sets each of the animators parameters to their corresponding State values
/// </summary>
function UpdateAnimator() {
	if (horizontalInput != 0) {
		if (horizontalInput < 0) flipX = true;
		else flipX = false;
	}

	_animator.SetBool("Grounded", isTouchingFloor);
	_animator.SetFloat("Speed", Mathf.Abs(currSpeed.x));
	_animator.SetFloat("vSpeed", currSpeed.y);
	//	_animator.SetBool("Running", BehaviorState.Running);
	//	_animator.SetBool("Dashing", BehaviorState.Dashing);
	_animator.SetBool("Crouching", Inputs.down.state == key_state.Hold);
	_animator.SetBool("LookingUp", Inputs.up.state == key_state.Hold);
	_animator.SetBool("WallClinging", !isTouchingFloor && isOnLeftWall || isOnRightWall);
	_animator.SetBool("Jetpacking", slow_fall);
	//_animator.SetBool("Diving", BehaviorState.Diving);
	//	_animator.SetBool("LadderClimbing", BehaviorState.LadderClimbing);
	//	_animator.SetFloat("LadderClimbingSpeed", BehaviorState.LadderClimbingSpeed);
	//	_animator.SetBool("FiringStop", BehaviorState.FiringStop);
	//_animator.SetBool("Firing", Inputs.fire.state == key_state.Down);
	//	_animator.SetInteger("FiringDirection", BehaviorState.FiringDirection);
	//	_animator.SetBool("MeleeAttacking", Inputs.grab.state == key_state.Hold);
	if (isOnRightWall) flipX = true;
	if (isOnLeftWall) flipX = false;
	if (flipX) transform.localScale.x = -1;
	else transform.localScale.x = 1;


}



function handleJumps() {
	jumpMax = jump_force;
	if (Inputs.jump.state == key_state.Down) {
		if (isTouchingFloor) {
			isStillJumping = true;
			currSpeed.y += (staticFrame.y + jumpMax);
		} else if (isOnLeftWall || isOnRightWall) {
			currSpeed.y = (staticFrame.y + jumpMax / 1.5);
			currSpeed.x += (isOnLeftWall) ? jumpMax : 0;
			currSpeed.x -= (isOnRightWall) ? jumpMax : 0;
			isStillJumping = true;
		}
    }
   	if (Inputs.jump.state == key_state.Up) {
		if (isStillJumping && currSpeed.y > 5) {
			currSpeed.y -= Mathf.Min(currSpeed.y / 2, jumpMax / 2);
		}
		isStillJumping = false;
   	}
	   slow_fall = (!isTouchingFloor && !isStillJumping && Inputs.jump.state == key_state.Hold);
}

var ground_speed = 0;
function handleHorizontal() {
	ground_speed = currSpeed.x - staticFrame.x;
	var ground_speed_magnitude = Mathf.Abs(ground_speed);
	speedMax = (((Inputs.down.state == key_state.None) ? runSpeed : runSpeed / 3) + Mathf.Abs(staticFrame.x));
	var dif = speedMax - ground_speed_magnitude;
	if (Mathf.Abs(horizontalInput) > 0.1) {
		if ((Mathf.Sign(horizontalInput) == Mathf.Sign(ground_speed))) {
			currSpeed.x += dif * (Time.fixedDeltaTime * accel) * Mathf.Sign(horizontalInput);
		} else {
			currSpeed.x += speedMax * (Time.fixedDeltaTime * accel) * Mathf.Sign(horizontalInput) * 2;
		}

	} else { //Stopping from no input
		if (ground_speed_magnitude > 4) {
			if (Inputs.down.state == key_state.Hold) {
				currSpeed.x -= (ground_speed * Time.fixedDeltaTime * accel) / 1.5;
			} else {
				currSpeed.x -= ground_speed * Time.fixedDeltaTime * accel * 4;
			}
		} else {
			//	print("stopped");
			currSpeed.x = staticFrame.x;
		}
	}
}

function getInputs() {
	if (Inputs.left.state == key_state.Hold) {
		horizontalInput = -1;
	} else if (Inputs.right.state == key_state.Hold) {
		horizontalInput = 1;
	} else {
		horizontalInput = 0;
	}
	//horizontalInput = movementJoy.position.x;
	if (Inputs.down.state == key_state.Hold) {
		verticalInput = -1;
	} else if (Inputs.up.state == key_state.Hold) {
		verticalInput = 1;
	} else {
		verticalInput = 0;
	}
}

function whatAmITouching() {
	isTouchingFloor = false;
	isOnLeftWall = false;
	isOnRightWall = false;

	var horizontal_look_dist = 1;
	var horizontal_center_offset = 2;
	var origin_2d = Vector2(transform.position.x, transform.position.y);
	var left_2d = Vector2((transform.position - transform.right * horizontal_look_dist).x, (transform.position - transform.right * horizontal_look_dist).y);
	var right_2d = Vector2((transform.position + transform.right * horizontal_look_dist).x, (transform.position + transform.right * horizontal_look_dist).y);
	var up_2d = Vector2((transform.position - transform.up * 1.5).x, (transform.position - transform.up * 1.5).y);

	right_ray = Physics2D.Linecast(origin_2d, right_2d, floor_layer_mask);
	if (right_ray.collider != null) {
		isOnRightWall = true;
		Debug.DrawLine(origin_2d, right_2d, Color.red, 2);
	}

	right_ray = Physics2D.Linecast(origin_2d + new Vector2(0, 1), right_2d + new Vector2(0, 1), floor_layer_mask);
	if (right_ray.collider != null) {
		isOnRightWall = true;
		Debug.DrawLine(origin_2d + new Vector2(0, 1), right_2d + new Vector2(0, 1), Color.red, 2);
	}

	left_ray = Physics2D.Linecast(origin_2d + new Vector2(0, 1), left_2d + new Vector2(0, 1), floor_layer_mask);
	if (left_ray.collider != null) {
		isOnLeftWall = true;
		Debug.DrawLine(origin_2d + new Vector2(0, 1), left_2d + new Vector2(0, 1), Color.yellow, 2);
	}

	left_ray = Physics2D.Linecast(origin_2d, left_2d, floor_layer_mask);
	if (left_ray.collider != null) {
		isOnLeftWall = true;
		Debug.DrawLine(origin_2d, left_2d, Color.yellow, 2);
	}

	floor_ray = Physics2D.Linecast(origin_2d + Vector2.left / 2, up_2d + Vector2.left / 2, floor_layer_mask);
	if (floor_ray.collider != null) {
		isTouchingFloor = true;
		Debug.DrawLine(origin_2d + Vector2.left / 2, up_2d + Vector2.left / 2, Color.blue, 2);
	}
	floor_ray = Physics2D.Linecast(origin_2d - Vector2.left / 2, up_2d - Vector2.left / 2, floor_layer_mask);
	if (floor_ray.collider != null) {
		isTouchingFloor = true;
		Debug.DrawLine(origin_2d - Vector2.left / 2, up_2d - Vector2.left / 2, Color.blue, 2);
	}
	isAirborn = (!isTouchingFloor && !isOnLeftWall && !isOnRightWall);
}


function OnCollisionStay2D(hit: Collision2D) {
	if (hit.gameObject.GetComponent.<Rigidbody2D>() && hit.rigidbody.mass * hit.rigidbody.velocity.magnitude > GetComponent.<Rigidbody2D>().mass * GetComponent.<Rigidbody2D>().velocity.magnitude) {
		setStaticFrame((hit.gameObject.GetComponent.<Rigidbody2D>().velocity));
	}
}


function setStaticFrame(outsideAction: Vector3) {
	//	staticFrame = outsideAction;
}



function applyForce(force: Vector2) {
	//Apply forces, but with positve y's'
	externalForces += (flipX) ? new Vector2(-force.x, force.y) : force;
}
function applyGlobalForce(force: Vector2) {
	externalForces += force;
}


function applyGravity() {
	if (currSpeed.y > -140) {
		if (((isOnLeftWall && Inputs.left.state == key_state.Hold) || (isOnRightWall && Inputs.right.state == key_state.Hold) || slow_fall) && currSpeed.y < -10) {
			currSpeed.y = currSpeed.y / 1.02;
			//apply gravityies
		} else {
			currSpeed.y -= gravity * Time.fixedDeltaTime;
		}
		//add extra gravity if user wants it
		//	if (Inputs.down.state == key_state.Hold) {
		//		currSpeed.y -= gravity * Time.fixedDeltaTime;
		//	}

	} else {
		//////////print(currSpeed.y);
	}
}
function ReceiveDamage(amount){
	health -= amount;
	if(health < 0){
		killPlayer();
	}
}

function killPlayer(){
	gameManager.SpawnPlayer(player_num);
	Destroy(this.gameObject);

}
function OnDestroy() {
	
}
